### PR TITLE
Finalize Rust representation for standard gates

### LINKED ISSUE #12566: Add remaining standard gates to rust representation
### What should we add?

After #12459 has merged there is now a native rust representation of the standard gate library. However, due to the size of #12459 not all gates were added to this rust representation. This issue tracks the migration of the remaining gates so that we have the complete standard gate library in rust. The remaining gates are:

```[tasklist]
### Tasks
- [x] C3XGate https://github.com/Qiskit/qiskit/pull/12659
- [x] C3SXGate https://github.com/Qiskit/qiskit/pull/12659
- [ ] C4XGate https://github.com/Qiskit/qiskit/pull/12659
- [x] DCXGate https://github.com/Qiskit/qiskit/pull/12644
- [x] CHGate https://github.com/Qiskit/qiskit/pull/12639
- [x] CPhaseGate https://github.com/Qiskit/qiskit/pull/12639
- [x] CRXGate https://github.com/Qiskit/qiskit/pull/12648
- [x] CRYGate https://github.com/Qiskit/qiskit/pull/12648
- [x] CRZGate https://github.com/Qiskit/qiskit/pull/12648
- [x] CSGate https://github.com/Qiskit/qiskit/pull/12639
- [x] CSdgGate https://github.com/Qiskit/qiskit/pull/12639
- [x] CSwapGate https://github.com/Qiskit/qiskit/pull/12639
- [x] CSXGate https://github.com/Qiskit/qiskit/pull/12639
- [x] CUGate https://github.com/Qiskit/qiskit/pull/12659
- [x] CU1Gate https://github.com/Qiskit/qiskit/pull/12659
- [x] CU3Gate https://github.com/Qiskit/qiskit/pull/12659
- [x] CCZGate https://github.com/Qiskit/qiskit/pull/12659
- [x] MSGate
- [x] PhaseGate
- [ ] RCCXGate https://github.com/Qiskit/qiskit/pull/12659
- [ ] RC3XGate https://github.com/Qiskit/qiskit/pull/12659
- [x] RGate https://github.com/Qiskit/qiskit/pull/12662 ~https://github.com/Qiskit/qiskit/pull/12507~
- [x] RXXGate https://github.com/Qiskit/qiskit/pull/12672
- [x] RYYGate https://github.com/Qiskit/qiskit/pull/12672
- [x] RZZGate https://github.com/Qiskit/qiskit/pull/12672
- [x] RZXGate https://github.com/Qiskit/qiskit/pull/12672
- [x] XXMinusYYGate https://github.com/Qiskit/qiskit/pull/12606
- [x] XXPlusYYGate https://github.com/Qiskit/qiskit/pull/12606
- [x] SGate https://github.com/Qiskit/qiskit/pull/12598
- [x] SdgGate https://github.com/Qiskit/qiskit/pull/12598
- [x] iSwapGate https://github.com/Qiskit/qiskit/pull/12598
- [x] SXGate https://github.com/Qiskit/qiskit/pull/12598
- [x] SXdgGate https://github.com/Qiskit/qiskit/pull/12598
- [x] TGate https://github.com/Qiskit/qiskit/pull/12598
- [x] TdgGate https://github.com/Qiskit/qiskit/pull/12598
- [x] U1Gate https://github.com/Qiskit/qiskit/pull/12572
- [x] U2Gate https://github.com/Qiskit/qiskit/pull/12572
- [x] U3Gate https://github.com/Qiskit/qiskit/pull/12572
```

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
* Added missing gate definitions and update rust equivalence tests accordingly. 
* Reordered gates in ascending number of qubits and alphabetical-ish order
* Reordered gate definitions and matrices to follow enum order for consistency
* Removed C4XGate from list of standard gates in rust

### Details and comments
This PR completes https://github.com/Qiskit/qiskit/issues/12566 🎉 

The proposed gate order is as follows (52 gates in total). Starting 1q gate list from `HGate` as suggested by @alexanderivrii.

```
pub enum StandardGate {
    // 0-qubits
    GlobalPhaseGate = 0,
    // 1-qubit
    HGate = 1,
    IGate = 2,
    XGate = 3,
    YGate = 4,
    ZGate = 5,
    PhaseGate = 6,
    RGate = 7,
    RXGate = 8,
    RYGate = 9,
    RZGate = 10,
    SGate = 11,
    SdgGate = 12,
    SXGate = 13,
    SXdgGate = 14,
    TGate = 15,
    TdgGate = 16,
    UGate = 17,
    U1Gate = 18,
    U2Gate = 19,
    U3Gate = 20,
    // 2-qubits
    CHGate = 21,
    CXGate = 22,
    CYGate = 23,
    CZGate = 24,
    DCXGate = 25,
    ECRGate = 26,
    SwapGate = 27,
    ISwapGate = 28,
    CPhaseGate = 29,
    CRXGate = 30,
    CRYGate = 31,
    CRZGate = 32,
    CSGate = 33,
    CSdgGate = 34,
    CSXGate = 35,
    CUGate = 36,
    CU1Gate = 37,
    CU3Gate = 38,
    RXXGate = 39,
    RYYGate = 40,
    RZZGate = 41,
    RZXGate = 42,
    XXMinusYYGate = 43,
    XXPlusYYGate = 44,
    // 3-qubits
    CCXGate = 45,
    CCZGate = 46,
    CSwapGate = 47,
    RCCXGate = 48,
    // 4-qubits
    C3XGate = 49,
    C3SXGate = 50,
    RC3XGate = 51,
}
```

### CODE REVIEW

File: crates/circuit/src/circuit_instruction.rs [hunk 1]
```diff
@@ -878,20 +878,37 @@ pub fn convert_py_to_operation_type(
...
+        let is_default_ctrl_state: bool = match py_op.getattr(py, intern!(py, "ctrl_state")) {
+            Ok(c_state) => match c_state.extract::<Option<i32>>(py) {
+                Ok(c_state_int) => match c_state_int {
+                    Some(c_int) => c_int == 1,
+                    None => true,
+                },
+                Err(_) => false,
+            },
+            Err(_) => false,
+        };
```
Comment (jakelishman) [inline crates/circuit/src/circuit_instruction.rs]: CCX's default control state is 3 - did we not include that in the standard gates?

Comment (ElePT) [reply crates/circuit/src/circuit_instruction.rs]: Aha, that makes sense, I wasn't thinking beyond 2q gates. I will generalize the logic.

Comment (ElePT) [reply crates/circuit/src/circuit_instruction.rs]: Addressed in #12659: [05fe760](https://github.com/Qiskit/qiskit/pull/12659/commits/05fe7608072dcb146189a0ca59c13864cafb272a)[05fe760](https://github.com/Qiskit/qiskit/pull/12659/commits/05fe7608072dcb146189a0ca59c13864cafb272a)

**Review by mtreinish — APPROVED** [review]
This LGTM, I personally probably wouldn't have bothered to reorder the match statements for the matrices and definitions because the order doesn't matter so much there. But being consistent with the canonical order makes sense.

**Pull request merged**